crypto/evp/p_lib.c

```diff
@@ -982,7 +982,7 @@ int EVP_PKEY_type(int type)
 
 int EVP_PKEY_get_id(const EVP_PKEY *pkey)
 {
-    return pkey->type;
+    return pkey->type > 0 ? pkey->type : NID_undef;
```

Comment:
The RETURN VALUES section for the `EVP_PKEY_get_id()` manual says this:

> **EVP_PKEY_get_base_id()**, **EVP_PKEY_get_id()** and **EVP_PKEY_type()** return a key type or **NID_undef** (equivalently **EVP_PKEY_NONE**) on error.

In practice, `pkey->type` is simply assigned `NID_undef` when it hasn't been assigned a value, so the "error" is really that it would be an error to try to use it for anything, as it's "undefined".  I personally disagree with how the RETURN VALUES documentation is written in this case.

However, as @slontis pointed out, there's the value `EVP_PKEY_KEYMGMT`, which is -1, which is there to signify that there is no NID associated with this `EVP_PKEY`, because it contains a purely provider based key.  That isn't an error per se, that key is supposed to be perfectly usable for diverse EVP operations...  but its identity must be retrieved by other means (as you figured out on your own).